Sources/Algorithms/Chunked.swift

```diff
@@ -246,3 +246,258 @@ extension Collection {
+where Base: RandomAccessCollection {
+  @inlinable
+  public func index(before i: Index) -> Index {
+    precondition(i > startIndex, "Advancing past start index")
+    
+    var offset = chunkCount
+    if i.baseRange.lowerBound == base.endIndex {
+      let remainder = base.count%chunkCount
+      if remainder != 0 {
+        offset = remainder
+      }
+    }
+    
+    let baseIdx = base.index(
+      i.baseRange.lowerBound, offsetBy: -offset,
+      limitedBy: base.startIndex
+    ) ?? base.startIndex
+    return Index(_baseRange: baseIdx..<i.baseRange.lowerBound)
+  }
+  
+  @inlinable
+  public func index(
+    _ i: Index, offsetBy offset: Int, limitedBy limit: Index
+  ) -> Index? {
+    guard offset != 0 else { return i }
+    guard limit != i else { return nil }
+    
+    if offset > 0 {
+      guard limit < i || distance(from: i, to: limit) >= offset else {
+        return nil
+      }
+      return offsetForward(i, offsetBy: offset)
+    } else {
+      guard limit > i || distance(from: i, to: limit) <= offset else {
+        return nil
+      }
+      return offsetBackward(i, offsetBy: offset)
+    }
+  }
+
+  @inlinable
+  public func index(_ i: Index, offsetBy distance: Int) -> Index {
+    guard distance != 0 else { return i }
+    
+    return distance > 0
+        ? offsetForward(i, offsetBy: distance)
+        : offsetBackward(i, offsetBy: distance)
+  }
+  
+  @usableFromInline
+  internal func offsetForward(_ i: Index, offsetBy distance: Int) -> Index {
+    return makeOffsetIndex(
+      from: i, baseBound: base.endIndex, distance: distance
+    )
+  }
+  
+  @usableFromInline
+  internal func offsetBackward(_ i: Index, offsetBy distance: Int) -> Index {
+    var idx = i
+    var distance = distance
+    // If we know that the last chunk is the only one that can possible
+    // have a variadic count. So in order to simplify and avoid another
+    // calculation of offsets(that is already done at `index(before:)`)
+    // we just move one position already so the index can be calculated
+    // since all remaining chunks have the same size.
+    if i.baseRange.lowerBound == base.endIndex {
+      formIndex(before: &idx)
+      distance += 1
+      // If the offset was simply one, we are done.
+      guard distance != 0 else {
+        return idx
+      }
+    }
+
+    return makeOffsetIndex(
+      from: idx, baseBound: base.startIndex, distance: distance
+    )
+  }
+  
+  // Helper to compute index(offsetBy:) index.
+  @inline(__always)
+  private func makeOffsetIndex(
+    from i: Index, baseBound: Base.Index, distance: Int
+  ) -> Index {
+    let baseStartIdx = base.index(
+      i.baseRange.lowerBound, offsetBy: distance * chunkCount,
+      limitedBy: baseBound
+    ) ?? baseBound
+    
+    let baseEndIdx = base.index(
+      i.baseRange.lowerBound, offsetBy: (distance + 1) * chunkCount,
+      limitedBy: base.endIndex
+    ) ?? base.endIndex
+    
+    return Index(_baseRange: baseStartIdx..<baseEndIdx)
+  }
```

Comment:
This is very tricky to get right, but ideally we would only traverse over the base indices between `i` and the output index once, not twice. We even shouldn't need to traverse the slice corresponding to `i.baseRange`, because we already know how big this chunk is.

Your code is completely correct and it matches the logarithmic performance characteristics of these methods, so there's no pressure to improve on this in this PR. 🙂

Reply:
Nice catch, just fixed! Thanks :)

Reply:
This looks like a good improvement, though this is still doing some unnecessary work (just less than before 😄). Offsetting an index forwards would ideally offset the upper bound of the index's base range, and not the lower bound. e.g. if you offset `Index(a..<b)` by `n`, then the lower bound of the output range is `(n - 1) * chunkSize` away from `b`, which is less work to compute (for non-random-access collections) than the index `n * chunkSize` away from `a`. And similarly, when offsetting an index backwards you would ideally first compute the upper bound of the output range and then the lower bound, rather than the other way around.

None of this matters much for random-access collections, but preferably on non-random-access collections these methods don't do more work than the naive solution of repeatedly calling `index(after:)` / `index(before:)`.

There doesn't seem to be a very easy way to fix this because you use `makeOffsetIndex` for both forward and backward offsets. Just to reiterate, this doesn't need to be perfect on the first go — there's a reason I still haven't finished my index offset logic for the `Product2` type 🙂 This stuff can be quite hairy.